File: concurrent_kata/mrugger/ruby/kata.sequence-iterator.rb
require "./kata.match-iterator"

class KataSequenceIterator

  @sequence
  @end_position
  @match_map

  @current_position

  def initialize(sequence, end_position, match_map)
    @sequence = sequence
    @end_position = end_position
    @current_position = sequence.length - 1
    @match_map = match_map
  end

  def each
    while (!is_done)
      yield self.next
    end
  end

  def next
    return nil if is_done
    iterator = KataMatchIterator.new(@sequence, @current_position)
    @current_position -= 1
    return iterator
  end

  def push_match(count)
    @match_map << { position: @current_position + 1, count: count }
  end


  private

  def is_done
    @current_position < @end_position
  end

end

File: concurrent_kata/mrugger/cplusplus/kata.sequence-iterator.cpp
#include <cassert>
#include <stdexcept>
#include "kata.sequence-iterator.h"
#include "kata.char-decrement-iterator.h"


SequenceIterator::SequenceIterator(char *block,
                                   const char *start,
                                   const char *end,
                                   const char *inner_loop_end,
                                   solution_stack *stack)
: _block(block),
  _cur_ptr(start),
  _end_ptr(end),
  _inner_loop_end(inner_loop_end),
  _stack(stack),
  _is_done(false)
{
  assert(start != NULL);
  assert(end != NULL);
  assert(start >= end);
  assert(inner_loop_end != NULL);
  assert(stack != NULL);
}


SequenceIterator::SequenceIterator(const char *start,
                                   const char *end,
                                   const char *inner_loop_end,
                                   solution_stack *stack)
: _cur_ptr(start),
  _end_ptr(end),
  _inner_loop_end(inner_loop_end),
  _stack(stack),
  _is_done(false)
{
  assert(start != NULL);
  assert(end != NULL);
  assert(start >= end);
  assert(inner_loop_end != NULL);
  assert(stack != NULL);
}


void SequenceIterator::for_each(std::function<int(CharIterable *)> lambda)
{
  std::unique_ptr<CharIterable> it(next());
  while (it.get() != NULL)
  {
    int count = lambda(it.get());
    push_match_count(count);
    it.reset(next());
  }
}


CharIterable *SequenceIterator::next(void)
{
  if (_is_done) return NULL;

  CharIterable *rv = new CharDecrementIterator(_cur_ptr--, _inner_loop_end);
  _is_done = _cur_ptr < _end_ptr;
  return rv;
}


void SequenceIterator::push_match_count(int count)
{
  if (count)
    _stack->push(solution_pair(_cur_ptr+1, count));
}
